vmm/src/device_manager.rs

```diff
@@ -653,22 +653,29 @@ impl DeviceRelocation for AddressManager {
             }
         }
 
-        let any_dev = pci_dev.as_any();
-        if let Some(virtio_pci_dev) = any_dev.downcast_ref::<VirtioPciDevice>() {
-            // Update the device_tree resources associated with the device
-            if let Some(node) = self
-                .device_tree
-                .lock()
-                .unwrap()
-                .get_mut(&virtio_pci_dev.id())
-            {
+        // Update the device_tree resources associated with the device
+        if let Some(id) = pci_dev.id() {
+            if let Some(node) = self.device_tree.lock().unwrap().get_mut(&id) {
                 let mut resource_updated = false;
                 for resource in node.resources.iter_mut() {
-                    if let Resource::MmioAddressRange { base, .. } = resource {
-                        if *base == old_base {
-                            *base = new_base;
-                            resource_updated = true;
-                            break;
+                    match region_type {
+                        PciBarRegionType::IoRegion => {
+                            if let Resource::PioAddressRange { base, .. } = resource {
```

Comment:
There is a 1:1 mapping between `PciBarRegionType::IoRegion` and `Resource::PioAddressRange`. An assertion could be used here to reinforce that.

Reply:
Yes I can check we don't end up with incompatible types. But I'd rather return an error rather that asserting, don't you think?

Reply:
Okay that makes sense :+1:

Reply:
Okay so I was updating the code when I realized this would break the intended behavior. During a BAR relocation, we know the type of BAR that was moved, therefore we only look for corresponding BAR types in the list of resources, and ultimately we check the address and size to identify the right one. So there's nothing to do here.

Reply:
One thing that might be interesting and that we could discuss for a follow up PR would be to introduce a new type of resource called `Resource::Bar{ id: u8, base: u64, size: usize }` as this would simplify some of the assumptions we have to make to retrieve the Pio or Mmio ranges corresponding to the BARs and making the code slightly more prone to errors.

Reply:
@sboeuf You are right. I omitted the context of the code, and now it makes sense to me.

As we are talking about handling the error case (where the expected resource is missing), do you think we need to replace the returning `Err` with an assertion?

Reply:
Let's leave it this way as it also covers the regular codepath as well.